lib/src/git.rs

```diff
@@ -361,9 +364,15 @@ pub fn fetch(
     fetch_options.proxy_options(proxy_options);
     let callbacks = callbacks.into_git();
     fetch_options.remote_callbacks(callbacks);
-    let refspec: &[&str] = &[];
+    if globs.iter().any(|g| g.contains(|c| ":^".contains(c))) {
+        return Err(GitFetchError::InvalidGlob);
+    }
+    let refspecs = globs
+        .iter()
+        .map(|glob| format!("refs/heads/{glob}:refs/heads/{glob}"))
```

Comment:
I thought the default, in the absence of a refspec, was akin to `+refs/heads/*:refs/remotes/REMOTENAME/*` (at least in git). Are there no ill effect pulling directly into `refs/heads/`?  At least there isn't the `+`, it will only go forward. Will the "BRANCHNAME@REMOTENAME" be advanced as well? Maybe including the output of `jj branch list` in the tests would be useful to visualize what is happening with and without `--branch`.

Reply:
> Are there no ill effect pulling directly into refs/heads/

Are you suggesting that I just use `let refspecs=globs` here? I was worried this could lead to strange behavior if there are slashes in the glob that could put the repository in a weird state. I personally use branches called `ig/something`, though maybe I shouldn't.

I'm also wondering whether it would be better to always include a +, i.e. `format!("+refs/heads/{glob}:refs/heads/{glob}")`. I'll see how that affects the tests. It would be nice to figure out what's the cause of the divergence bug that's marked in the tests (when old positions of the branch aren't abandoned).

> Maybe including the output of jj branch list in the tests would be useful to visualize what is happening with and without --branch.

It's worth trying.

Reply:
> I'm also wondering whether it would be better to always include a +...

This fixed the bug! Yay! Only one bug left now (and I think it's essentially an instance of #922).

Reply:
I also included some `jj branch list` output. At a glance, it's as expected.

Reply:
> Are you suggesting that I just use `let refspecs=globs` here?

I was rather wondering why the refspec wasn't `format!("+refs/heads/{glob}:refs/remotes/{remote}/{glob}")` (provided `remote` is the remote name, I haven't checked the code) so that `*` as `glob` would have the same behaviour as without any glob.

Reply:
Here is a test in a colocated git repo which fails with the current refspec https://gist.github.com/samueltardieu/99ecbc3ef90c60a92fb31aa118d4454a as the local `rem1` branch is lost when fetching.

The correct result seems to be obtained with `format!("+refs/heads/{glob}:refs/remotes/{remote_name}/{glob}")`.

Reply:
Thank you very much! I added your tests. I also changed one of my tests to create conflicting branches, and that uncovered a bunch of other problems with my original refspec, all fixed by the change you suggested. (If you are interested, try changing the refspec back to what it was and see how the tests change).